### PR TITLE
TST: GH30999 Add match=msg to all but two pytest.raises in tests/io

### PR DESCRIPTION
**Pull request by moink**
This pull request partially addresses xref #30999 to remove bare `pytest.raises` by adding `match=msg`. It doesn't close that issue as I have only addressed test modules in the pandas/tests/io/ directory.

I thought this was going to be relatively small because there were only 25 such instances, but it turned out some of them were in methods that were reused several times, so this ended up larger than I expected.

There were a couple of cases where I found the structure of the tests quite confusing, with an internal function definition and two nested asserts. This structure also made it difficult to add the assertion about the error message. I restructured the tests to be, in my opinion, more straightforward.

There's one file, pandas/tests/io/test_sql.py, with two instances of `pytest.raises` in tests which were skipped on my machine and which I couldn't figure out how to run, so those are left as is.

- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/parser/test_python_parser_only.py [hunk 1]
```diff
@@ -249,19 +249,15 @@ def test_multi_char_sep_quotes(python_parser_only, quoting):
...
-    def fail_read():
+    if quoting == csv.QUOTE_NONE:
+        msg = "Expected 2 fields in line 3, saw 3"
         with pytest.raises(ParserError, match=msg):
             parser.read_csv(StringIO(data), quoting=quoting, **kwargs)
-
-    if quoting == csv.QUOTE_NONE:
-        # We expect no match, so there should be an assertion
-        # error out of the inner context manager.
-        with pytest.raises(AssertionError):
```
Comment (jreback) [inline pandas/tests/io/parser/test_python_parser_only.py:261]: thanks, this was some really old testing code

File: pandas/tests/io/test_fsspec.py [hunk 2]
```diff
@@ -223,11 +221,12 @@ def test_s3_parquet(s3_resource, s3so):
     tm.assert_equal(df1, df2)
 
 
+@td.skip_if_installed("fsspec")
```
Comment (jreback) [inline pandas/tests/io/test_fsspec.py:224]: extra skip here

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. one comment. pls merge master and ping on green.

**Comment by jreback** [comment]
> There were a couple of cases where I found the structure of the tests quite confusing, with an internal function definition and two nested asserts. This structure also made it difficult to add the assertion about the error message. I restructured the tests to be, in my opinion, more straightforward.

yep your changes are fine. general restructuring of tests should be a dedicated PR, but moderinzation like this is ok.

File: pandas/tests/io/test_fsspec.py (hunk 2 above)
Comment (moink) [reply pandas/tests/io/test_fsspec.py:224]: @jreback fixed this, merged master, green except for Travis CI

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @moink

**Pull request merged**